jwt/keyset.go

```diff
@@ -0,0 +1,211 @@
+	}, nil
+}
+
+// VerifySignature parses the given JWT, verifies its signature using JWKS keys, and returns
+// the claims in its payload. The given JWT must be of the JWS compact serialization form.
+func (ks JSONWebKeySet) VerifySignature(ctx context.Context, token string) (map[string]interface{}, error) {
+	payload, err := ks.remoteJWKS.VerifySignature(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal payload into a set of all received claims
+	allClaims := map[string]interface{}{}
+	if err := json.Unmarshal(payload, &allClaims); err != nil {
+		return nil, err
+	}
+
+	return allClaims, nil
+}
+
+// NewStaticKeySet returns a KeySet that verifies JWT signatures using the given publicKeys.
+func NewStaticKeySet(publicKeys []crypto.PublicKey) (KeySet, error) {
+	return StaticKeySet{
+		publicKeys: publicKeys,
+	}, nil
+}
+
+// VerifySignature parses the given JWT, verifies its signature using local public keys, and
+// returns the claims in its payload. The given JWT must be of the JWS compact serialization form.
+func (ks StaticKeySet) VerifySignature(_ context.Context, token string) (map[string]interface{}, error) {
+	parsedJWT, err := jwt.ParseSigned(token)
+	if err != nil {
+		return nil, err
+	}
+
+	var valid bool
+	allClaims := map[string]interface{}{}
+	for _, key := range ks.publicKeys {
+		if err := parsedJWT.Claims(key, &allClaims); err == nil {
+			valid = true
+			break
+		}
+	}
+	if !valid {
+		return nil, errors.New("no known key successfully validated the token signature")
+	}
+
+	return allClaims, nil
+}
+
+// ParsePublicKeyPEM is used to parse RSA and ECDSA public keys from PEMs. The given
+// data must be of PEM-encoded x509 certificate or PKIX public key forms. It returns
+// an *rsa.PublicKey or *ecdsa.PublicKey.
+func ParsePublicKeyPEM(data []byte) (crypto.PublicKey, error) {
+	block, data := pem.Decode(data)
+	if block != nil {
+		var rawKey interface{}
+		var err error
+		if rawKey, err = x509.ParsePKIXPublicKey(block.Bytes); err != nil {
+			if cert, err := x509.ParseCertificate(block.Bytes); err == nil {
+				rawKey = cert.PublicKey
+			} else {
+				return nil, err
+			}
+		}
+
+		if rsaPublicKey, ok := rawKey.(*rsa.PublicKey); ok {
+			return rsaPublicKey, nil
+		}
+		if ecPublicKey, ok := rawKey.(*ecdsa.PublicKey); ok {
+			return ecPublicKey, nil
+		}
+	}
+
+	return nil, errors.New("data does not contain any valid RSA or ECDSA public keys")
+}
+
+// createCAContext returns a context with a custom TLS client that's configured with the root
+// certificates from caPEM. If no certificates are configured, the original context is returned.
+func createCAContext(ctx context.Context, caPEM string) (context.Context, error) {
+	if caPEM == "" {
+		return ctx, nil
+	}
+
+	certPool := x509.NewCertPool()
+	if ok := certPool.AppendCertsFromPEM([]byte(caPEM)); !ok {
+		return nil, errors.New("could not parse CA PEM value successfully")
+	}
+
+	tr := cleanhttp.DefaultPooledTransport()
+	if certPool != nil {
```

Comment:
I don't think `certPool` can ever be `nil` if it is instantiated by `x509.NewCertPool()`.

Reply:
That's true. Removed in 5d29657.